site/guides/effective-go-krm-function.md

```diff
@@ -0,0 +1,212 @@
+```go
+func (r *YourFunction) Run(context *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+	// namespaceScoped contains only namespace scoped resources  
+    namespaceScoped := objects.Where(func(o *fn.KubeObject) bool { return o.IsNamespaceScoped() }) 
+    // clusterScoped contains only cluster scoped resources
+    clusterScoped := objects.Where(func(o *fn.KubeObject) bool { return o.IsClusterScoped() })
+    // customDeployment contains all resources of Kind "CustomDeployment", in Group "fn.kpt.dev" with any Versions.  
+    customDeployment := objects.Where(fn.IsGVK("fn.kpt.dev", "", "CustomDeployment") })
+    // excluded contains all resources except namespace objects
+    excluded := objects.WhereNot(fn.IsGVK("v1", "", "Namespace") })
+}
+```
+
+### Read and write a field spec path of a KRM resource
+
+Like [unstructured.Unstructured], `fn.KubeObject` (and `fn.SubObject`) provides a series of methods to 
+let you read and write different resources types.
+
+```go
+func (r *YourFunction) Run(context *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+	// Get first deployment object.
+	deployment := items.Where(fn.IsGVK("apps", "v1", "Deployment")).Where(func(o *fn.KubeObject) bool{return o.GetName() == "nginx"})[0]
+	// Get the int value from deployment `spec.replicas`
+	replicas := deployment.NestedInt64OrDie("spec", "replicas")
+	fn.Logf("replicas is %v\n", replicas)
+    // Get the boolean value from deployment `spec.paused`
+    paused := deployment.NestedBoolOrDie("spec", "paused")
+    fn.Logf("paused is %v\n", paused)
+    // Update strategy from Recreate to RollingUpdate.
+    if strategy := obj.NestedStringOrDie("spec", "strategy", "type"); strategy == "Recreate" {
+    	obj.SetNestedStringOrDie("RollingUpdate", "spec", "strategy", "type")
+    }
+}
+```
+
+Besides the [unstructured.Unstructured] style, you can also run functions on each sub-field as a `fn.SubObject`  
+```go
+func (r *YourFunction) Run(context *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+    // Get first deployment object.
+	deployment := items.Where(fn.IsGVK("apps", "v1", "Deployment")).Where(func(o *fn.KubeObject) bool{return o.GetName() == "nginx"})[0]
+    // Get a spec as a `SubObject`
+	spec := deployment.GetMap("spec")
+    // Get integer from SubObject spec
+    replicas = spec.GetInt("replicas")
+    fn.Logf("replicas is %v\n", replicas)
+    // Get the SubObject from another SubObject
+    nodeSelector := spec.GetMap("template").GetMap("spec").GetMap("nodeSelector")
+    if nodeSelector.GetString("disktype") != "ssd" {
+       nodeSelector.SetNestedStringOrDie("ssd", "disktype")
+    }
+}
+```
+
+### Copy `KubeObject` to a typed struct
+
+If you already have some struct to define a KRM resource (like `corev1.ConfigMap`), you can switch the `KubeObject` 
+to the other type via `As`
```

Comment:
I would add an explicit warning that this copies the object and thus you have to copy it back